examples/dynamic-forall.cpp

```diff
@@ -77,28 +77,25 @@ int main(int argc, char *argv[])
 
 //----------------------------------------------------------------------------//
 
-  std::cout << "\n Running C-style vector addition...\n";
-
-  // _cstyle_vector_add_start
-  for (int i = 0; i < N; ++i) {
-    c[i] = a[i] + b[i];
-  }
-  // _cstyle_vector_add_end
-
-  checkResult(c, N);
-//printResult(c, N);
-
-
-//----------------------------------------------------------------------------//
-// Example of dynamic policy selection for forall
-//----------------------------------------------------------------------------//
+  std::cout << "\n Running dynamic forall vector addition and reductions...\n";
 
+  int sum = 0;
+  using VAL_INT_SUM = RAJA::expt::ValOp<int, RAJA::operators::plus>;
```

Comment:
Why do we need this? The way the machinery works on the back, it would be just as easy to hand the lambda a reference to an int.  Is this to prevent users from using a different operator on the target or something?  Either way shouldn't hold this PR up, but it would probably help with the verbosity of these reductions a significant amount.

Reply:
We've recently modified the new-reducer interface to accept only these `ValOp` objects. This allows the operations within the lambda to be consistent, e.g. `valopobj.minloc(array, loc);` rather than the macros for loc reductions we were using before. It also allows us to ensure some type and operator safety among the possible reduction objects, e.g. a sum reducer cannot call `min()`. The `ValOp` objects take references to the original `int` or location, so nothing functionally changes for the user outside of the RAJA lambda. Yes, it's more verbose and somewhat ugly, but if nvcc were to allow auto deduction of types in lambda parameters, then we could get rid of some of this verbosity.

Reply:
Ok, I see where you're coming from. On the `*loc` objects, it makes sense to me to have a helper type for those that has an `<op>loc` or similar on it, but why not allow the regular reference for the single-component types? It's sometimes useful to do different operations in the body than at the actual reduction.  Usually that's for things like getting a per-thread value, like "max iterations computed by one thread" or similar, but it does occasionally come up.

To be explicit, I think the `ValOp` types are a good idea like the strongly-typed indexes, but it seems like something that should be optional rather than required.  Also, I would expect to be able to get it by doing `RAJA::expt::Reduce<RAJA::operators::plus, int>::arg_type`, or `decltype(RAJA::expt::Reduce<RAJA::operators::plus>(&sum))::arg_type`, or similar so the declaration for the type used to pass in the reducer provides the appropriate `ValOp` type.

Reply:
I think these are good topics, I can capture the comments here as an issue for further discussion. But I think the topic here may be out of the scope for this PR?

Reply:
Sounds right @artv3.